CacheStorageRecordInformation is not isolated copied when passed across threads

szewai opened the pull request:
#### 9d8de91ee2cbfa9731b1b8171818b193c254daf7
<pre>
CacheStorageRecordInformation is not isolated copied when passed across threads
<a href="https://bugs.webkit.org/show_bug.cgi?id=286393">https://bugs.webkit.org/show_bug.cgi?id=286393</a>
<a href="https://rdar.apple.com/108372325">rdar://108372325</a>

Reviewed by Chris Dumez.

CacheStorageRecordInformation only has isolatedCopy function for rvalue reference, but in
CacheStorageDiskStore::readRecordsInternal(), isolatedCopy() is not invoked on rvalue reference. In this case,
CrossThreadCopier will fall back to use copy constructor, and CacheStorageRecordInformation is not properly copied. To
fix it, add CacheStorageRecordInformation::isolatedCopy() for lvalue reference and make sure CrossThreadCopier will
invoke that.

Also, to make the code more robust, now types involved in crossThreadCopy() are required to provide isolatedCopy()
function -- CrossThreadCopier will no longer use default copy constructor. This helps catch a few issues where objects
are not properly copied due to missing isolatedCopy function.

* Source/WTF/wtf/CrossThreadCopier.h:
* Source/WTF/wtf/MediaTime.cpp:
(WTF::MediaTime::isolatedCopy const):
* Source/WTF/wtf/MediaTime.h:
* Source/WebCore/Modules/cookie-store/CookieChangeSubscription.h:
(WebCore::CookieChangeSubscription::isolatedCopy const):
(WebCore::CookieChangeSubscription::isolatedCopy):
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamConstructionParameters.h:
(WebCore::WebTransportBidirectionalStreamConstructionParameters::isolatedCopy):
* Source/WebCore/contentextensions/ContentExtensionsBackend.h:
(WebCore::ContentExtensions::ContentExtensionsBackend::isolatedCopy):
(WebCore::ContentExtensions::ContentExtensionsBackend::ContentExtensionsBackend):
* Source/WebCore/fileapi/URLKeepingBlobAlive.h:
* Source/WebCore/platform/ContentType.h:
(WebCore::ContentType::isolatedCopy const):
(WebCore::ContentType::isolatedCopy):
* Source/WebCore/platform/graphics/VideoPlaybackQualityMetrics.h:
(WebCore::VideoPlaybackQualityMetrics::isolatedCopy const):
* Source/WebCore/platform/mediastream/PhotoCapabilities.h:
(WebCore::PhotoCapabilities::isolatedCopy const):
* Source/WebCore/platform/mediastream/PhotoSettings.h:
(WebCore::PhotoSettings::isolatedCopy const):
* Source/WebCore/platform/network/OrganizationStorageAccessPromptQuirk.h:
(WebCore::OrganizationStorageAccessPromptQuirk::isolatedCopy const):
(WebCore::OrganizationStorageAccessPromptQuirk::isolatedCopy):
* Source/WebKit/NetworkProcess/storage/CacheStorageRecord.cpp:
(WebKit::CacheStorageRecordInformation::isolatedCopy):
(WebKit::CacheStorageRecordInformation::isolatedCopy const):
* Source/WebKit/NetworkProcess/storage/CacheStorageRecord.h:
* Source/WebKit/Shared/RTCNetwork.cpp:
(WebKit::RTCNetwork::isolatedCopy const):
* Source/WebKit/Shared/RTCNetwork.h:
(WebKit::RTC::Network::IPAddress::isolatedCopy const):
(WebKit::RTC::Network::InterfaceAddress::isolatedCopy const):
* Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
(WebKit::PlatformExtensionCapabilityGrants::isolatedCopy const):

Canonical link: <a href="https://commits.webkit.org/289347@main">https://commits.webkit.org/289347@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/571c80fe0e71a7048fc454fc431efe966a1eb854

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86570 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6082 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40900 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91419 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37307 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/88619 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6344 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14134 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66986 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24752 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89573 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4812 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78386 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47303 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4613 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32712 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36424 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/79357 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75104 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33596 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93300 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/85348 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13728 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9942 "Found 1 new test failure: fast/webgpu/type-checker-array-without-argument.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75764 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13929 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74227 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74955 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18444 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19235 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17627 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/6512 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13751 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/19007 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/107836 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13489 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25951 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16933 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15273 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/storage/CacheStorageRecord.h:
```diff
@@ -37,6 +37,7 @@ class CacheStorageRecordInformation {
     CacheStorageRecordInformation(NetworkCache::Key&&, double insertionTime, uint64_t identifier, uint64_t updateResponseCounter, uint64_t size, URL&&, bool hasVaryStar, HashMap<String, String>&& varyHeaders);
     void updateVaryHeaders(const WebCore::ResourceRequest&, const WebCore::ResourceResponse::CrossThreadData&);
     CacheStorageRecordInformation isolatedCopy() &&;
+    CacheStorageRecordInformation isolatedCopy() const &;
```
We need to fix the root cause in crossThreadCopier(). Otherwise it is bound to bite us again. Finding the bug is great and the fix is good but I think we should address the foot-gun in the same PR. I don't think we need to minimize the patch size here since this will not be cherry-picked.

I worry we may fail to follow-up to address the root cause otherwise. Also, fixing `crossThreadCopier()` may identify other bugs that would be very valuable too.

cdumez commented on Source/WTF/wtf/CrossThreadCopier.h:
```diff
@@ -92,12 +92,21 @@ template<typename T>
     static constexpr bool IsNeeded = HasIsolatedCopy<T>::value;
-    template<typename U> static auto copy(U&& value) -> Type
+
+    template<typename U> static Type copy(const U& value)
```
How would this change help catch the issue? It seems to me like it will still implicitly copy if no isolatedCopy() is available? Shouldn't we require an `isolatedCopy()` to be available if `crossThreadCopy()` gets called on a non-built-in type?

szewai replied:
This will help catch the issue where a Type provides `isolatedCopy() &&` but `not isolatedCopy() const &`. There will be a compile error as when we invoke `crossThreadCopy()` is on `const T&`.

> Shouldn't we require an isolatedCopy() to be available if crossThreadCopy() gets called on a non-built-in type?

My understanding is our behavior so far is to invoke `isolatedCopy()` when it's available, else fall back to copy constructor; so if a type need deep copy we need to add `isolatedCopy()` function for it.
This is not related to the bug because the type has `isolatedCopy()` function. 

Do you think we should require `isolatedCopy()` on all non-builtin types? Might be a bigger change but I can investigate if it's preferred

cdumez replied:
> Do you think we should require isolatedCopy() on all non-builtin types? Might be a bigger change but I can investigate if it's preferred

Can you take a look to see how big of a change? I suspect it wouldn't be *that* big but not 100% sure.
I still feel this API is very error-prone. Given how hard these kind of bugs are to investigate, I think it is worth the effort.

cdumez commented on Source/WTF/wtf/MediaTime.h:
```diff
@@ -136,6 +136,7 @@ class WTF_EXPORT_PRIVATE MediaTime final {
     };
 
     MediaTime toTimeScale(uint32_t, RoundingFlags = RoundingFlags::HalfAwayFromZero) const;
+    MediaTime isolatedCopy() const &;
```
No need for `&`

cdumez commented on Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamConstructionParameters.h:
```diff
@@ -41,6 +41,7 @@ struct WebTransportBidirectionalStreamConstructionParameters {
     WEBCORE_EXPORT WebTransportBidirectionalStreamConstructionParameters(WebTransportBidirectionalStreamConstructionParameters&&);
     WEBCORE_EXPORT WebTransportBidirectionalStreamConstructionParameters& operator=(WebTransportBidirectionalStreamConstructionParameters&&);
     WEBCORE_EXPORT ~WebTransportBidirectionalStreamConstructionParameters();
+    WebTransportBidirectionalStreamConstructionParameters isolatedCopy() && { return { identifier, Ref { sink } }; }
```
`&&` doesn't seem needed / helpful, also can be `const`.

szewai replied:
The struct only has move constructor, cannot construct result with `WebTransportBidirectionalStreamConstructionParameters { identifier,  sink }`

cdumez commented on Source/WebCore/contentextensions/ContentExtensionsBackend.h:
```diff
@@ -84,7 +85,14 @@ class ContentExtensionsBackend {
+    ContentExtensionsBackend isolatedCopy() && { return { crossThreadCopy(WTFMove(m_contentExtensions)) }; }
+
 private:
+    ContentExtensionsBackend(UncheckedKeyHashMap<String, Ref<ContentExtension>>&& contentExtensions)
```
`explicit`

cdumez commented on Source/WebCore/fileapi/URLKeepingBlobAlive.h:
```diff
@@ -62,3 +63,19 @@ class URLKeepingBlobAlive {
 } // namespace WebCore
+
+namespace WTF {
+template<> struct CrossThreadCopierBase<false, false, WebCore::URLKeepingBlobAlive> {
```
I don't fully understand why we need this template specialization. Likely related to the fact that the class is not copyable but I still don't get why. It's a bit unfortunate to have to do this.

szewai replied:
Right, this is not needed any more (needed for previous iteration)

cdumez commented on Source/WebCore/platform/graphics/VideoPlaybackQualityMetrics.h:
```diff
@@ -45,6 +45,17 @@ struct VideoPlaybackQualityMetrics {
         return *this;
     }
+
+    VideoPlaybackQualityMetrics isolatedCopy() const &
```
`&` is not needed.

cdumez replied:
Can we just `return *this;`

cdumez commented on Source/WebCore/platform/mediastream/PhotoCapabilities.h:
```diff
@@ -35,6 +35,16 @@ enum class FillLightMode : uint8_t;
 enum class RedEyeReduction : uint8_t;
 
 struct PhotoCapabilities {
+    PhotoCapabilities isolatedCopy() const &
```
`&` is not needed.

cdumez commented on Source/WebCore/platform/mediastream/PhotoCapabilities.h:
```diff
@@ -35,6 +35,16 @@ enum class FillLightMode : uint8_t;
 struct PhotoCapabilities {
+    PhotoCapabilities isolatedCopy() const &
+    {
+        return {
```
Can we just `return *this;` ?

cdumez commented on Source/WebCore/platform/mediastream/PhotoSettings.h:
```diff
@@ -33,6 +33,16 @@ namespace WebCore {
 enum class FillLightMode : uint8_t;
 
 struct PhotoSettings {
+    PhotoSettings isolatedCopy() const &
```
ditto.

cdumez commented on Source/WebKit/Shared/RTCNetwork.h:
```diff
@@ -60,6 +60,10 @@ struct IPAddress {
     {
     }
 
+    IPAddress isolatedCopy() const & {
```
`&` is not needed and `{` is on the wrong line.

cdumez commented on Source/WebKit/Shared/RTCNetwork.h:
```diff
@@ -74,6 +78,9 @@ struct InterfaceAddress {
     }
 
     rtc::InterfaceAddress rtcAddress() const;
+    InterfaceAddress isolatedCopy() const & {
```
ditto.

cdumez commented on Source/WebKit/Shared/RTCNetwork.h:
```diff
@@ -104,6 +111,7 @@ struct RTCNetwork {
 
     RTCNetwork() = default;
     explicit RTCNetwork(Vector<char>&& name, Vector<char>&& description, IPAddress prefix, int prefixLength, int type, uint16_t id, int preference, bool active, bool ignored, int scopeID, Vector<InterfaceAddress>&& ips);
+    RTCNetwork isolatedCopy() const &;
```
`&` is not needed.

cdumez reviewed (approved):
Awesome patch, you fixed so many bugs!

szewai committed:
CacheStorageRecordInformation is not isolated copied when passed across threads
https://bugs.webkit.org/show_bug.cgi?id=286393
rdar://108372325

Reviewed by Chris Dumez.

CacheStorageRecordInformation only has isolatedCopy function for rvalue reference, but in
CacheStorageDiskStore::readRecordsInternal(), isolatedCopy() is not invoked on rvalue reference. In this case,
CrossThreadCopier will fall back to use copy constructor, and CacheStorageRecordInformation is not properly copied. To
fix it, add CacheStorageRecordInformation::isolatedCopy() for lvalue reference and make sure CrossThreadCopier will
invoke that.

Also, to make the code more robust, now types involved in crossThreadCopy() are required to provide isolatedCopy()
function -- CrossThreadCopier will no longer use default copy constructor. This helps catch a few issues where objects
are not properly copied due to missing isolatedCopy function.

* Source/WTF/wtf/CrossThreadCopier.h:
* Source/WTF/wtf/MediaTime.cpp:
(WTF::MediaTime::isolatedCopy const):
* Source/WTF/wtf/MediaTime.h:
* Source/WebCore/Modules/cookie-store/CookieChangeSubscription.h:
(WebCore::CookieChangeSubscription::isolatedCopy const):
(WebCore::CookieChangeSubscription::isolatedCopy):
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamConstructionParameters.h:
(WebCore::WebTransportBidirectionalStreamConstructionParameters::isolatedCopy):
* Source/WebCore/contentextensions/ContentExtensionsBackend.h:
(WebCore::ContentExtensions::ContentExtensionsBackend::isolatedCopy):
(WebCore::ContentExtensions::ContentExtensionsBackend::ContentExtensionsBackend):
* Source/WebCore/fileapi/URLKeepingBlobAlive.h:
* Source/WebCore/platform/ContentType.h:
(WebCore::ContentType::isolatedCopy const):
(WebCore::ContentType::isolatedCopy):
* Source/WebCore/platform/graphics/VideoPlaybackQualityMetrics.h:
(WebCore::VideoPlaybackQualityMetrics::isolatedCopy const):
* Source/WebCore/platform/mediastream/PhotoCapabilities.h:
(WebCore::PhotoCapabilities::isolatedCopy const):
* Source/WebCore/platform/mediastream/PhotoSettings.h:
(WebCore::PhotoSettings::isolatedCopy const):
* Source/WebCore/platform/network/OrganizationStorageAccessPromptQuirk.h:
(WebCore::OrganizationStorageAccessPromptQuirk::isolatedCopy const):
(WebCore::OrganizationStorageAccessPromptQuirk::isolatedCopy):
* Source/WebKit/NetworkProcess/storage/CacheStorageRecord.cpp:
(WebKit::CacheStorageRecordInformation::isolatedCopy):
(WebKit::CacheStorageRecordInformation::isolatedCopy const):
* Source/WebKit/NetworkProcess/storage/CacheStorageRecord.h:
* Source/WebKit/Shared/RTCNetwork.cpp:
(WebKit::RTCNetwork::isolatedCopy const):
* Source/WebKit/Shared/RTCNetwork.h:
(WebKit::RTC::Network::IPAddress::isolatedCopy const):
(WebKit::RTC::Network::InterfaceAddress::isolatedCopy const):
* Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
(WebKit::PlatformExtensionCapabilityGrants::isolatedCopy const):

Canonical link: https://commits.webkit.org/289347@main
